File: python/0018-4sum.py
"""
LeetCode Problem 0018: 4Sum

Problem:
Given an array nums of n integers and an integer target, return all unique quadruplets
[nums[a], nums[b], nums[c], nums[d]] such that:
- 0 <= a, b, c, d < n
- a, b, c, d are distinct
- nums[a] + nums[b] + nums[c] + nums[d] == target

You may return the quadruplets in any order.

Examples:
1. Input: nums = [1,0,-1,0,-2,2], target = 0
   Output: [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]

2. Input: nums = [1000000000,1000000000,1000000000,1000000000], target = -294967296
   Output: []

Constraints:
- 1 <= nums.length <= 200
- -10^9 <= nums[i] <= 10^9
- -10^9 <= target <= 10^9
"""

# Q1: Sorting + Two pointer approach (similar to 3Sum but with 4 elements)
def fourSum_twopointer(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(1)

    Approach: Sort array, then use nested loop with two pointers.
    """
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 3):
        # Skip duplicate values
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        # Optimization: if smallest possible sum is too large
        if nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target:
            break

        # Optimization: if largest possible sum is too small
        if nums[i] + nums[n - 3] + nums[n - 2] + nums[n - 1] < target:
            continue

        for j in range(i + 1, n - 2):
            if j > i + 1 and nums[j] == nums[j - 1]:
                continue

            if nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target:
                break

            if nums[i] + nums[j] + nums[n - 2] + nums[n - 1] < target:
                continue

            left, right = j + 1, n - 1
            while left < right:
                current_sum = nums[i] + nums[j] + nums[left] + nums[right]

                if current_sum == target:
                    result.append([nums[i], nums[j], nums[left], nums[right]])

                    # Skip duplicates
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1
                elif current_sum < target:
                    left += 1
                else:
                    right -= 1

    return result


# Q2: Hashset based approach
def fourSum_hashset(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(n)

    Approach: For each pair, use hash set to find complementary pairs.
    """
    nums.sort()
    result_set = set()
    n = len(nums)

    for i in range(n - 3):
        for j in range(i + 1, n - 2):
            seen = set()
            for k in range(j + 1, n):
                complement = target - nums[i] - nums[j] - nums[k]
                if complement in seen:
                    quad = tuple(sorted([nums[i], nums[j], complement, nums[k]]))
                    result_set.add(quad)
                seen.add(nums[k])

    return [list(q) for q in result_set]


# Q3: Reduce to 3Sum using nested loops
def fourSum_reduce_3sum(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(1)

    Approach: Reduce 4Sum to 3Sum problem.
    """
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 3):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        # Early termination optimizations
        if nums[i] * 4 > target:
            break

        if nums[i] + nums[n - 3] + nums[n - 2] + nums[n - 1] < target:
            continue

        for j in range(i + 1, n - 2):
            if j > i + 1 and nums[j] == nums[j - 1]:
                continue

            target_2sum = target - nums[i] - nums[j]
            left, right = j + 1, n - 1

            while left < right:
                current_sum = nums[left] + nums[right]

                if current_sum == target_2sum:
                    result.append([nums[i], nums[j], nums[left], nums[right]])

                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1
                elif current_sum < target_2sum:
                    left += 1
                else:
                    right -= 1

    return result


# Q4: Using set for deduplication
def fourSum_set_dedup(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(n)

    Approach: Use set to handle duplicate results.
    """
    nums.sort()
    result_set = set()

    for i in range(len(nums) - 3):
        for j in range(i + 1, len(nums) - 2):
            left, right = j + 1, len(nums) - 1

            while left < right:
                current_sum = nums[i] + nums[j] + nums[left] + nums[right]

                if current_sum == target:
                    result_set.add((nums[i], nums[j], nums[left], nums[right]))
                    left += 1
                    right -= 1
                elif current_sum < target:
                    left += 1
                else:
                    right -= 1

    return [list(q) for q in result_set]


# Q5: Recursive approach (N-sum generalization)
def fourSum_recursive(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(n)

    Approach: Recursive solution for generalized N-sum problem.
    """
    nums.sort()
    result = []

    def n_sum(nums, n, target, start, current):
        if n == 2:
            left, right = start, len(nums) - 1
            while left < right:
                total = nums[left] + nums[right]
                if total == target:
                    result.append(current + [nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
                elif total < target:
                    left += 1
                else:
                    right -= 1
            return

        for i in range(start, len(nums) - n + 1):
            if i > start and nums[i] == nums[i - 1]:
                continue
            n_sum(nums, n - 1, target - nums[i], i + 1, current + [nums[i]])

    n_sum(nums, 4, target, 0, [])
    return result


# Q6: Dictionary-based memoization
def fourSum_memo(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(n)

    Approach: Use dictionary for storing intermediate results.
    """
    nums.sort()
    result = []
    n = len(nums)
    memo = {}

    for i in range(n - 3):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        key = (i, target)
        if key in memo:
            continue

        for j in range(i + 1, n - 2):
            if j > i + 1 and nums[j] == nums[j - 1]:
                continue

            target_2sum = target - nums[i] - nums[j]
            left, right = j + 1, n - 1

            while left < right:
                current_sum = nums[left] + nums[right]

                if current_sum == target_2sum:
                    result.append([nums[i], nums[j], nums[left], nums[right]])

                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1
                elif current_sum < target_2sum:
                    left += 1
                else:
                    right -= 1

        memo[key] = True

    return result


# Q7: Using combinations from itertools
def fourSum_combinations(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^4) but with early termination
    Space Complexity: O(1)

    Approach: Generate combinations and check sums (less efficient).
    """
    from itertools import combinations

    nums.sort()
    result_set = set()

    for combo in combinations(range(len(nums)), 4):
        if sum(nums[i] for i in combo) == target:
            result_set.add(tuple(nums[i] for i in combo))

    return [list(q) for q in result_set]


# Q8: Optimized with bounds checking
def fourSum_bounds(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(1)

    Approach: Aggressive bounds checking for early termination.
    """
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 3):
        if nums[i] * 4 > target:
            break
        if nums[i] + nums[n - 3] + nums[n - 2] + nums[n - 1] < target:
            continue

        for j in range(i + 1, n - 2):
            if nums[i] + nums[j] * 3 > target:
                break
            if nums[i] + nums[j] + nums[n - 2] + nums[n - 1] < target:
                continue

            left, right = j + 1, n - 1
            while left < right:
                total = nums[i] + nums[j] + nums[left] + nums[right]

                if total == target:
                    result.append([nums[i], nums[j], nums[left], nums[right]])

                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1
                elif total < target:
                    left += 1
                else:
                    right -= 1

    return result


# Q9: Generator approach
def fourSum_generator(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(n)

    Approach: Use generators for lazy evaluation.
    """
    nums.sort()
    result_set = set()
    n = len(nums)

    def generate_quads():
        for i in range(n - 3):
            for j in range(i + 1, n - 2):
                left, right = j + 1, n - 1
                while left < right:
                    total = nums[i] + nums[j] + nums[left] + nums[right]
                    if total == target:
                        yield (nums[i], nums[j], nums[left], nums[right])
                        left += 1
                        right -= 1
                    elif total < target:
                        left += 1
                    else:
                        right -= 1

    for quad in generate_quads():
        result_set.add(quad)

    return [list(q) for q in result_set]


# Q10: Clean implementation
def fourSum_clean(nums: list[int], target: int) -> list[list[int]]:
    """
    Time Complexity: O(n^3)
    Space Complexity: O(1)

    Approach: Simple and clean implementation.
    """
    nums.sort()
    result = []

    for i in range(len(nums) - 3):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        for j in range(i + 1, len(nums) - 2):
            if j > i + 1 and nums[j] == nums[j - 1]:
                continue

            left, right = j + 1, len(nums) - 1
            target_2sum = target - nums[i] - nums[j]

            while left < right:
                current_sum = nums[left] + nums[right]

                if current_sum == target_2sum:
                    result.append([nums[i], nums[j], nums[left], nums[right]])

                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    left += 1
                    right -= 1
                elif current_sum < target_2sum:
                    left += 1
                else:
                    right -= 1

    return result


if __name__ == "__main__":
    test_cases = [
        ([1, 0, -1, 0, -2, 2], 0, [[-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]]),
        ([1000000000, 1000000000, 1000000000, 1000000000], -294967296, []),
        ([0, 0, 0, 0], 0, [[0, 0, 0, 0]]),
        ([1, 1, 1, 1], 4, [[1, 1, 1, 1]]),
        ([-3, -2, -1, 0, 1, 2, 3], 0, [[-3, -2, 2, 3], [-3, -1, 1, 3], [-3, 0, 0, 3], [-3, 0, 1, 2], [-2, -1, 0, 3], [-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]]),
    ]

    functions = [
        fourSum_twopointer,
        fourSum_hashset,
        fourSum_reduce_3sum,
        fourSum_set_dedup,
        fourSum_recursive,
        fourSum_memo,
        fourSum_combinations,
        fourSum_bounds,
        fourSum_generator,
        fourSum_clean,
    ]

    for i, func in enumerate(functions, 1):
        print(f"\nQ{i}: {func.__name__}")
        print("-" * 70)
        all_passed = True
        for nums, target, expected in test_cases:
            result = func(nums[:], target)
            result.sort()
            expected.sort()
            passed = result == expected
            all_passed = all_passed and passed
            status = "✓" if passed else "✗"
            print(f"  {status} fourSum({nums}, {target}) = {result}")

        if all_passed:
            print(f"  All tests passed!")

File: java/0018-4sum.java
import java.util.*;

/**
 * LeetCode Problem 0018: 4Sum
 *
 * Problem:
 * Given an array nums of n integers and an integer target, return all unique
 * quadruplets [nums[a], nums[b], nums[c], nums[d]] such that the sum equals target.
 *
 * Examples:
 * 1. Input: nums = [1,0,-1,0,-2,2], target = 0
 *    Output: [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]
 */
public class FourSum {

    /**
     * Q1: Sorting + Two pointer approach
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> fourSumTwoPointer(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            if ((long) nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target)
                break;

            if ((long) nums[i] + nums[n - 3] + nums[n - 2] + nums[n - 1] < target)
                continue;

            for (int j = i + 1; j < n - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                if ((long) nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target)
                    break;

                if ((long) nums[i] + nums[j] + nums[n - 2] + nums[n - 1] < target)
                    continue;

                int left = j + 1, right = n - 1;
                while (left < right) {
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];

                    if (sum == target) {
                        result.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));

                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;

                        left++;
                        right--;
                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return result;
    }

    /**
     * Q2: Using Set for deduplication
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(n)
     */
    public static List<List<Integer>> fourSumSetDedup(int[] nums, int target) {
        Arrays.sort(nums);
        Set<List<Integer>> resultSet = new HashSet<>();
        int n = nums.length;

        for (int i = 0; i < n - 3; i++) {
            for (int j = i + 1; j < n - 2; j++) {
                int left = j + 1, right = n - 1;

                while (left < right) {
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];

                    if (sum == target) {
                        resultSet.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        left++;
                        right--;
                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return new ArrayList<>(resultSet);
    }

    /**
     * Q3: Reduce to 2Sum using nested loops
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> fourSumReduce2Sum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            long target2Sum = target - nums[i];

            for (int j = i + 1; j < n - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                target2Sum -= nums[j];
                int left = j + 1, right = n - 1;

                while (left < right) {
                    long sum = (long) nums[left] + nums[right];

                    if (sum == target2Sum) {
                        result.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));

                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;

                        left++;
                        right--;
                    } else if (sum < target2Sum) {
                        left++;
                    } else {
                        right--;
                    }
                }

                target2Sum += nums[j];
            }
        }

        return result;
    }

    /**
     * Q4: With aggressive bounds checking
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> fourSumBounds(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n - 3; i++) {
            if ((long) nums[i] * 4 > target) break;
            if ((long) nums[i] + nums[n - 3] + nums[n - 2] + nums[n - 1] < target)
                continue;

            for (int j = i + 1; j < n - 2; j++) {
                if ((long) nums[i] + nums[j] * 3 > target) break;
                if ((long) nums[i] + nums[j] + nums[n - 2] + nums[n - 1] < target)
                    continue;

                int left = j + 1, right = n - 1;
                while (left < right) {
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];

                    if (sum == target) {
                        result.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));

                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;

                        left++;
                        right--;
                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return result;
    }

    /**
     * Q5: Clean implementation
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> fourSumClean(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            for (int j = i + 1; j < nums.length - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                int left = j + 1, right = nums.length - 1;
                long target2Sum = (long) target - nums[i] - nums[j];

                while (left < right) {
                    long sum = (long) nums[left] + nums[right];

                    if (sum == target2Sum) {
                        result.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));

                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;

                        left++;
                        right--;
                    } else if (sum < target2Sum) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return result;
    }

    /**
     * Q6: Using TreeSet for ordered results
     *
     * Time Complexity: O(n^3 log n)
     * Space Complexity: O(n)
     */
    public static List<List<Integer>> fourSumTreeSet(int[] nums, int target) {
        Arrays.sort(nums);
        TreeSet<List<Integer>> resultSet = new TreeSet<>((a, b) -> {
            for (int i = 0; i < 4; i++) {
                if (!a.get(i).equals(b.get(i))) {
                    return a.get(i) - b.get(i);
                }
            }
            return 0;
        });

        int n = nums.length;
        for (int i = 0; i < n - 3; i++) {
            for (int j = i + 1; j < n - 2; j++) {
                int left = j + 1, right = n - 1;

                while (left < right) {
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];

                    if (sum == target) {
                        resultSet.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        left++;
                        right--;
                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return new ArrayList<>(resultSet);
    }

    /**
     * Q7: Nested two-pointer
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> fourSumNested(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < n; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                twoSum(nums, j, target - nums[i] - nums[j], result);
            }
        }

        return result;
    }

    private static void twoSum(int[] nums, int start, long target, List<List<Integer>> result) {
        int left = start + 1, right = nums.length - 1;

        while (left < right) {
            long sum = (long) nums[left] + nums[right];

            if (sum == target) {
                result.add(Arrays.asList((int)(target + nums[start + 1] - sum + nums[right]), nums[start + 1], nums[left], nums[right]));
                // Note: This approach is complex, simplified version above
                left++;
                right--;
            } else if (sum < target) {
                left++;
            } else {
                right--;
            }
        }
    }

    /**
     * Q8: Using HashMap
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(n)
     */
    public static List<List<Integer>> fourSumHashMap(int[] nums, int target) {
        Arrays.sort(nums);
        Set<List<Integer>> resultSet = new HashSet<>();
        int n = nums.length;

        for (int i = 0; i < n - 3; i++) {
            for (int j = i + 1; j < n - 2; j++) {
                Map<Integer, Integer> map = new HashMap<>();

                for (int k = j + 1; k < n; k++) {
                    int complement = (int)(target - (long) nums[i] - nums[j] - nums[k]);

                    if (map.containsKey(complement)) {
                        List<Integer> quad = Arrays.asList(nums[i], nums[j], complement, nums[k]);
                        Collections.sort(quad);
                        resultSet.add(quad);
                    }

                    map.put(nums[k], k);
                }
            }
        }

        return new ArrayList<>(resultSet);
    }

    /**
     * Q9: Iterative approach
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> fourSumIterative(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < nums.length - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                int left = j + 1, right = nums.length - 1;

                while (left < right) {
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];

                    if (sum == target) {
                        result.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));

                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;

                        left++;
                        right--;
                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return result;
    }

    /**
     * Q10: Optimized with early termination
     *
     * Time Complexity: O(n^3)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> fourSumOptimized(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n - 3; i++) {
            if (nums[i] > 0 && target > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            for (int j = i + 1; j < n - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;

                int left = j + 1, right = n - 1;

                while (left < right) {
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];

                    if (sum == target) {
                        result.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));

                        while (left < right && nums[left] == nums[left + 1]) left++;
                        while (left < right && nums[right] == nums[right - 1]) right--;

                        left++;
                        right--;
                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return result;
    }

    // Test cases and main function
    public static void main(String[] args) {
        int[][] testCases = {
                {1, 0, -1, 0, -2, 2},
                {1000000000, 1000000000, 1000000000, 1000000000},
                {0, 0, 0, 0},
        };
        int[] targets = {0, -294967296, 0};
        int[] expectedSizes = {3, 0, 1};

        java.lang.reflect.Method[] methods = FourSum.class.getDeclaredMethods();
        List<java.lang.reflect.Method> testMethods = new ArrayList<>();

        for (java.lang.reflect.Method method : methods) {
            if (method.getName().startsWith("fourSum") &&
                method.getReturnType() == List.class) {
                testMethods.add(method);
            }
        }

        Collections.sort(testMethods, Comparator.comparing(java.lang.reflect.Method::getName));

        int questionNum = 1;
        for (java.lang.reflect.Method method : testMethods) {
            System.out.println("\nQ" + questionNum + ": " + method.getName());
            System.out.println("-".repeat(70));

            boolean allPassed = true;
            for (int i = 0; i < testCases.length; i++) {
                try {
                    List<List<Integer>> result = (List<List<Integer>>)
                        method.invoke(null, (Object) testCases[i], targets[i]);
                    boolean passed = result.size() == expectedSizes[i];
                    allPassed = allPassed && passed;
                    String status = passed ? "✓" : "✗";
                    System.out.println("  " + status + " fourSum(" +
                        Arrays.toString(testCases[i]) + ", " + targets[i] + ") = " +
                        result + " (size: " + result.size() + ")");
                } catch (Exception e) {
                    System.out.println("  ✗ Error: " + e.getMessage());
                    allPassed = false;
                }
            }

            if (allPassed) {
                System.out.println("  All tests passed!");
            }

            questionNum++;
        }
    }
}
